### PR TITLE
esp8266: Change from FAT to littlefs v2 as default filesystem.

### PR DESCRIPTION
**Pull request by dpgeorge**
This PR changes the esp8266 boards to use littlefs v2 as the filesystem, rather than FAT.  Since the esp8266 doesn't expose the filesystem to the PC over USB there's no strong reason to keep it as FAT.  Littlefs is smaller in code size, is more efficient in use of flash to store data, is resilient over power failure, and using it saves about 4k of heap RAM, which can now be used for other things.

This is a backwards incompatible change because all existing esp8266 boards will need to update their filesystem after installing new firmware (ie backup old files, install firmware, restore files to new filesystem).

As part of this PR the memory layout of the default board has changed.  It now allocates all 1M of memory-mapped flash to the firmware, so the filesystem area starts at the 2M point.  This is done to allow more frozen bytecode to be stored in the 1M of memory-mapped flash.  This requires an esp8266 module with 2M or more of flash to work, so a new board called `GENERIC_1M` is added which has the old memory-mapping (but still changed to use littlefs).

In summary there are now 3 esp8266 board definitions:
- `GENERIC_512K` for 512k modules, doesn't have a filesystem
- `GENERIC_1M` for 1M modules, 572k for firmware+frozen code, 396k for filesystem (littlefs)
- `GENERIC_2M` for 2M (or greater) modules, 968k for firmware+frozen code, 1M+ for filesystem (littlefs)

The plan is to have all 3 available for download at https://micropython.org/download

### CODE REVIEW

**Comment by tve** [comment]
Tangential question: would you entertain switching to littlefs by default on the esp32 also?

**Comment by dpgeorge** [comment]
> would you entertain switching to littlefs by default on the esp32 also?

Ah, yes, it would be good to do that at the same time.  Note that the esp32 already enables `VfsLfs2`, so one can already opt-in to littlefs.  So the change would be to make littlefs the default for new systems, but old systems that use FAT would continue to work.

**Comment by dpgeorge** [comment]
See #5811 for corresponding esp32 change.

**Comment by robert-hh** [comment]
Can a SD card still be used with a FAT filesystem?

**Comment by dpgeorge** [comment]
> Can a SD card still be used with a FAT filesystem?

Good point, no it cannot.  So... probably the firmware needs both FAT and littlefs drivers built in.  For modules with 1M flash it's pretty tight so maybe just FAT.  For 2M+ modules it could have both, but if you're not using FAT then that's wasting about 20k of flash.  Alternative would be to provide multiple builds, with FAT, LFS2, FAT+LFS2.

**Comment by robert-hh** [comment]
If I look at the available ESP8266 modules that have the pins for SD accessible, then most of them have 4MB flash or more. Therefore it seems acceptable to me, that modules with 1MB or less have littlefs only, and for the others both littlefs and FAT is supported as default. People with different needs can still build their own firmware.

**Comment by dpgeorge** [comment]
> Therefore it seems acceptable to me, that modules with 1MB or less have littlefs only, and for the others both littlefs and FAT is supported as default.

Sounds good to me, and I've pushed a commit to do this.

Note that enabling FAT takes about 21000 bytes of flash and 512 bytes RAM (for a static working long-filename buffer).

**Comment by dpgeorge** [comment]
Now merged, with a few minor changes to comments in the ld files, updated `ports/esp8266/makeimg.py` file, and enabled building of new `GENERIC_1M` board on Travis.

I tested this but there may still be something I missed, hopefully the community can test it further.

**Comment by Ozturk470** [comment]
Ben bunların ne olduğunu çok anlamadım öyle engeller diye açtım virüsleri
de olayın virüs olmadığını huyum kurusun geç anlıyorum.

4 Nis 2020 Cmt, saat 09:01 tarihinde Damien George <notifications@github.com>
şunu yazdı:

**Comment by Ozturk470** [comment]
Devam ederse ciddi bir destek için devletten yardım isteyeceğim

4 Nis 2020 Cmt, saat 09:03 tarihinde Direk mesaj Eşkiya <
eskiyadirekmesaj@gmail.com> şunu yazdı:

**Comment by Ozturk470** [comment]
Eğitim amaçlı ücretsiz kurslar var galiba bu kurumlardan öğrenmek istiyorum
ücretli olanlara yazdım 1600 tl dediler

4 Nis 2020 Cmt, saat 09:05 tarihinde Direk mesaj Eşkiya <
eskiyadirekmesaj@gmail.com> şunu yazdı:

**Comment by Ozturk470** [comment]
Telefon Üstünden suç işlese birisi fark etmiceZ resmen

4 Nis 2020 Cmt, saat 09:06 tarihinde Direk mesaj Eşkiya <
eskiyadirekmesaj@gmail.com> şunu yazdı:

**Comment by mcauser** [comment]
[The ESP8266 docs tutorial](http://docs.micropython.org/en/latest/esp8266/tutorial/filesystem.html) could use an update. They still say FAT is the default, and there is no mention there are 3 board defs with different configs. Found [this page](http://docs.micropython.org/en/latest/reference/filesystem.html) with more up to date infos. Maybe add a link?
